server/datastore/mysql/software.go

```diff
@@ -200,25 +200,41 @@ func deleteUninstalledHostSoftwareDB(
 	currentMap map[string]fleet.Software,
 	incomingMap map[string]fleet.Software,
 ) error {
-	var deletesHostSoftware []interface{}
-	deletesHostSoftware = append(deletesHostSoftware, hostID)
-
+	var deletesHostSoftware []uint
 	for currentKey, curSw := range currentMap {
 		if _, ok := incomingMap[currentKey]; !ok {
 			deletesHostSoftware = append(deletesHostSoftware, curSw.ID)
 		}
 	}
-	if len(deletesHostSoftware) <= 1 {
+	if len(deletesHostSoftware) == 0 {
 		return nil
 	}
-	sql := fmt.Sprintf(
-		`DELETE FROM host_software WHERE host_id = ? AND software_id IN (%s)`,
-		strings.TrimSuffix(strings.Repeat("?,", len(deletesHostSoftware)-1), ","),
-	)
-	if _, err := tx.ExecContext(ctx, sql, deletesHostSoftware...); err != nil {
+
+	stmt := `DELETE FROM host_software WHERE host_id = ? AND software_id IN (?);`
+	stmt, args, err := sqlx.In(stmt, hostID, deletesHostSoftware)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "build delete host software query")
+	}
+	if _, err := tx.ExecContext(ctx, stmt, args...); err != nil {
 		return ctxerr.Wrap(ctx, err, "delete host software")
 	}
 
+	// Cleanup the software table when no more hosts have the deleted host_software
+	// table entries.
+	// Otherwise the software will be listed by ds.ListSoftware but ds.SoftwareByID,
+	// ds.CountHosts and ds.ListHosts will return a *notFoundError error for such
+	// software.
+	stmt = `DELETE FROM software WHERE id IN (?) AND id NOT IN (
+		SELECT DISTINCT software_id FROM host_software WHERE software_id IN (?)
+	);`
```

Comment:
I _think_ a `NOT EXISTS (SELECT 1 FROM host_software WHERE sofware_id IN (?))` would be faster, as it wouldn't have to build the whole list of `software_id` before running the `NOT IN` condition?

Reply:
Mhm I don't follow. Could you share the full query you have in mind?

What I want is to delete all `software` entries that have been deleted from *this* host AND that are not present in `host_software` anymore. (Basically the current host was the only one with the software installed and has now removed it, so we don't need it in `software` anymore.)

Reply:
Sure thing, what I had in mind was something like that:

```
DELETE FROM software sw
WHERE sw.id IN (?) AND 
NOT EXISTS (
   SELECT 1 FROM host_software hsw WHERE hsw.software_id = sw.id
)
```

Reply:
Ah, yes, the query is much clearer and probably faster (though the explain shows it's somewhat similar).
```
mysql> EXPLAIN EXTENDED DELETE FROM software WHERE id IN (1,2,3) AND NOT EXISTS (SELECT 1 FROM host_software hsw WHERE hsw.software_id = software.id)
    -> ;
+----+--------------------+----------+------------+-------+---------------------------+---------------------------+---------+-------------------+------+----------+-------------+
| id | select_type        | table    | partitions | type  | possible_keys             | key                       | key_len | ref               | rows | filtered | Extra       |
+----+--------------------+----------+------------+-------+---------------------------+---------------------------+---------+-------------------+------+----------+-------------+
|  1 | DELETE             | software | NULL       | range | PRIMARY                   | PRIMARY                   | 8       | const             |    3 |   100.00 | Using where |
|  2 | DEPENDENT SUBQUERY | hsw      | NULL       | ref   | host_software_software_fk | host_software_software_fk | 8       | fleet.software.id |    1 |   100.00 | Using index |
+----+--------------------+----------+------------+-------+---------------------------+---------------------------+---------+-------------------+------+----------+-------------+
2 rows in set, 2 warnings (0.01 sec)

mysql> EXPLAIN DELETE FROM software WHERE id IN (1,2,3) AND id NOT IN (SELECT DISTINCT software_id FROM host_software WHERE software_id IN (1,2,3));
+----+--------------------+---------------+------------+----------------+---------------------------+---------------------------+---------+-------+------+----------+--------------------------+
| id | select_type        | table         | partitions | type           | possible_keys             | key                       | key_len | ref   | rows | filtered | Extra                    |
+----+--------------------+---------------+------------+----------------+---------------------------+---------------------------+---------+-------+------+----------+--------------------------+
|  1 | DELETE             | software      | NULL       | range          | PRIMARY                   | PRIMARY                   | 8       | const |    3 |   100.00 | Using where              |
|  2 | DEPENDENT SUBQUERY | host_software | NULL       | index_subquery | host_software_software_fk | host_software_software_fk | 8       | func  |    1 |   100.00 | Using where; Using index |
+----+--------------------+---------------+------------+----------------+---------------------------+---------------------------+---------+-------+------+----------+--------------------------+
2 rows in set (0.01 sec)
```

Anyways I've changed the query. Thanks!